pallets/file-system/src/utils.rs

```diff
@@ -0,0 +1,112 @@
+use frame_support::{ensure, pallet_prelude::DispatchResult, sp_runtime::BoundedVec, traits::Get};
+
+use crate::{
+    pallet,
+    types::{FileLocation, Fingerprint, MultiAddress, StorageRequestMetadata, StorageUnit},
+    Error, Pallet, StorageRequestExpirations, StorageRequests,
+};
+
+macro_rules! expect_or_err {
+    ($optional:expr, $error_msg:expr, $error_type:path) => {
+        match $optional {
+            Some(value) => value,
+            None => {
+                #[cfg(test)]
+                unreachable!($error_msg);
+
+                #[allow(unreachable_code)]
+                {
+                    Err($error_type)?
+                }
+            }
+        }
+    };
+}
+
+impl<T> Pallet<T>
+where
+    T: pallet::Config,
+{
+    pub fn do_request_storage(
+        location: FileLocation<T>,
+        fingerprint: Fingerprint<T>,
+        size: StorageUnit<T>,
+        user_multiaddr: MultiAddress<T>,
+    ) -> DispatchResult {
+        // TODO: Check user funds and lock them for the storage request.
+        // TODO: Check storage capacity of chosen MSP (when we support MSPs)
+        // TODO: Return error if the file is already stored and overwrite is false.
+        let current_block_number = <frame_system::Pallet<T>>::block_number();
+        let file_metadata = StorageRequestMetadata::<T> {
+            requested_at: current_block_number,
+            fingerprint,
+            size,
+            user_multiaddr,
+            bsps_volunteered: BoundedVec::default(),
+            bsps_confirmed: BoundedVec::default(),
+        };
+
+        // Check that storage request is not already registered.
+        ensure!(
+            !<StorageRequests<T>>::contains_key(&location),
+            Error::<T>::StorageRequestAlreadyRegistered
+        );
+
+        // Register storage request.
+        <StorageRequests<T>>::insert(&location, file_metadata);
+
+        // TODO: Maybe worth it to loop a few times until we find a free slot. (this should never fail and so a loop is required)
```

Comment:
Instead, we can have a `StorageValue` that holds the number of the current expiration block (`CurrentExpBlock`). The minimum expiration number of blocks should be the constant `StorageRequestTtl`, so when a new storage request is generated, you append it in `StorageReqeustExpiration[expiration_block]` where `expiation_block = max(StorageRequestTtl, CurrentExpBlock)`. That way, in O(1) you get the next available slot. When a block's vector is filled, you increment `CurrentExpBlock`, and then if `CurrentExpBlock < StorageRequestTtl`, then you update `CurrentExpBlock = StorageRequestTtl`.